src/main/java/com/multicraft/MultiCraftCommandExecutor.java

```diff
@@ -228,98 +168,72 @@ public boolean onCommand(CommandSender sender, Command cmd, String label, String
+				CopyHandler.getInstance().markStartPosition(p, startLocation);
 
-				Location startLocation = p.getTargetBlock((HashSet<Byte>) null, 32).getLocation();
-				if (p.getWorld().getBlockAt(startLocation).getType() == Material.AIR) {
-					p.sendMessage("Please find a closer position");
-					break;
-				}
-				copyHandler.setCopyLoc1(p, startLocation);
 				p.sendMessage("The first position has been marked.");
 				return true;
 			}
 			case "copyloc2": {
-				if (!p.isOp()) {
-					p.sendMessage("This command requires op status.");
+				if (playerRequirementNotMet(p, Player::isOp) || playerRequirementNotMet(p, playerInCreativeMode)) {
 					break;
 				}
 
-				if (p.getGameMode() != GameMode.CREATIVE) {
-					p.sendMessage("This command requires creative mode.");
+				Location endLocation = Commands.getPlayerTargetLocation(p, 16, true);
+				if (!CopyHandler.getInstance().markEndPosition(p, endLocation)) {
+					p.sendMessage("You cannot mark an end location without beginning a valid region session.");
 					break;
 				}
 
-				Location endLocation = p.getTargetBlock((HashSet<Byte>) null, 32).getLocation();
-				if (p.getWorld().getBlockAt(endLocation).getType() == Material.AIR) {
-					p.sendMessage("Please find a closer position");
-					break;
-				}
-				copyHandler.setCopyLoc2(p, endLocation);
 				p.sendMessage("The second position has been marked.");
 				return true;
 			}
 			case "mpaste": {
-				if (!p.isOp()) {
-					p.sendMessage("This command requires op status.");
-					break;
-				}
-
-				if (p.getGameMode() != GameMode.CREATIVE) {
-					p.sendMessage("This command requires creative mode.");
-					break;
-				}
-
-				String copyArgs = copyHandler.getCopyArgs(p);
-				if (copyArgs == null) {
-					p.sendMessage("Please ensure both copy locations are marked.");
+				String copyArgs = CopyHandler.getInstance().getCopyArgs(p);
+				if (playerRequirementNotMet(p, Player::isOp) || playerRequirementNotMet(p, playerInCreativeMode) || copyArgs.isEmpty()) {
 					break;
 				}
 
-				Location pasteLocation = p.getTargetBlock((HashSet<Byte>) null, 16).getLocation();
-				if (p.getWorld().getBlockAt(pasteLocation).getType() == Material.AIR) {
-					p.sendMessage("Please find a closer position");
-					break;
-				}
+				Location pasteLocation = Commands.getPlayerTargetLocation(p, 16, true);
+				String pasteArgs = pasteLocation.getX() + " " + pasteLocation.getY() + " " + pasteLocation.getZ() + " ";
 
-				String pasteArgs = pasteLocation.getX() + " " + (pasteLocation.getY() + 1) + " "
-						+ pasteLocation.getZ() + " ";
-				p.performCommand("clone " + copyArgs + pasteArgs + "masked");
-				return true;
+				return p.performCommand("clone " + copyArgs + pasteArgs + "masked");
 			}
 			default:
 				break;
 		}
 		p.sendMessage("No MultiCraft commands executed.");
 		return false;
 	}
+
+	private static boolean playerRequirementNotMet(Player p, Function<Player, Boolean> requirement) {
```

Comment:
I'm not sure if this extra wrapper function is necessary, or I think it should say which requirement was not met.

Reply:
That makes sense, I can make this into individual functions.